File: blogi_projekti/API/Controllers/UsersController.cs
// UsersController

using API.CustomExceptions;
using API.Dtos;
using API.Interfaces;
using AutoMapper;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace API.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    // Voimme injektoida IMapper-tyyppisenä _mapperin mihin tahansa controlleriin
    public class UsersController(IUserService _userService, IMapper _mapper) : ControllerBase
    {
        [HttpGet("account")]
        [Authorize]
        public async Task<ActionResult<UserDto>> GetAccount()
        {
            var idClaim = User.Claims.First(c => c.Type == "sub");
            var id = int.Parse(idClaim.Value);

            var user = await _userService.GetAccount(id);
            if (user == null)
            {
                return Forbid();
            }

            return Ok(_mapper.Map<UserDto>(user));
        }

        [HttpGet]
        [Authorize(Policy = "RequireAdminRole")]
        public async Task<ActionResult<IEnumerable<UserDto>>> GetAll()
        {
            var users = await _userService.GetAll();

            return Ok(_mapper.Map<IEnumerable<UserDto>>(users));
        }

        [HttpPost("login")]
        public async Task<ActionResult<LoginRes>> Login([FromBody] LoginReq request)
        {
            try
            {
                var token = await _userService.Login(request.UserName, request.Password);
                return Ok(new LoginRes { Token = token });
            }
            catch (NotFoundException)
            {
                return Unauthorized("invalid username or password");
            }
            catch (Exception e)
            {
                return Problem(
                    title: "Login failed",
                    detail: e.Message,
                    statusCode: StatusCodes.Status500InternalServerError
                );
            }
        }

        [HttpPost("register")]
        public async Task<ActionResult<RegisterRes>> Register(RegisterReq requestData)
        {
            try
            {
                var user = await _userService.Register(requestData.UserName, requestData.Password);
                return new RegisterRes
                {
                    Id = user.Id,
                    UserName = user.UserName,
                    Role = user.Role,
                };
            }
            // toistaiseksi missään ei vielä heitetä tätä omaa poikkeusta
            // otetaan se käyttöön myöhemmin
            catch (UserRegistrationException e)
            {
                return Problem(
                    title: "Error creating user",
                    detail: e.Message,
                    statusCode: StatusCodes.Status400BadRequest
                );
            }
            catch (Exception e)
            {
                return Problem(
                    title: "Error creating user",
                    detail: e.Message,
                    statusCode: StatusCodes.Status500InternalServerError
                );
            }
        }

        [HttpGet("account/rewards")]
        // tähän routeen ei pääse, jos Xp-avain puuttuu jwt-tokenista
        // tai sen arvo on alle 1000
        [Authorize(Policy = "Require1000Xp")]
        public ActionResult<string> GetRewards()
        {
            return Ok("");
        }
    }
}

File: blogi_projekti_python/controllers/users_controller.py
from typing import Annotated, List

from fastapi import APIRouter, Depends, HTTPException

import dtos.login
import dtos.register
import dtos.users
from custom_exceptions.not_found_exception import NotFoundException
from dependencies import require_admin, require_xp
from factories.services import user_service_factory
from services.abc_user_service import ABCUserService

router = APIRouter(prefix="/api/users", tags=["users"])

UserServ = Annotated[ABCUserService, Depends(user_service_factory)]


@router.post("/register")
async def register(
    request_data: dtos.register.RegisterReq, _user_service: UserServ
) -> dtos.register.RegisterRes:
    user = _user_service.register(request_data.username, request_data.password)
    return dtos.register.RegisterRes(
        id=user.Id,
        username=user.UserName,
        role=user.Role,
    )


@router.post("/login")
async def login(
    request_data: dtos.login.LoginReq, _user_service: UserServ
) -> dtos.login.LoginRes:
    try:
        token = _user_service.login(request_data.username, request_data.password)
        return dtos.login.LoginRes(token=token)
    except NotFoundException:
        raise HTTPException(status_code=404, detail="Username or password incorrect")
    except Exception as error:
        raise HTTPException(status_code=500, detail=str(error))


@router.get("/", dependencies=[Depends(require_admin)])
async def get_all_users(_user_service: UserServ) -> List[dtos.users.UserDto]:
    try:
        users = _user_service.get_all()

        # return [dtos.users.UserDto.model_validate(user) for user in users]
        return users
    except Exception as error:
        raise HTTPException(status_code=500, detail=str(error))


@router.get("/account/rewards", dependencies=[Depends(require_xp(1000))])
async def get_user_rewards():
    return "hello"
